File: types/buz_hash_boundary_checker.go
package types

import (
	"github.com/attic-labs/noms/Godeps/_workspace/src/github.com/attic-labs/buzhash"
	"github.com/attic-labs/noms/d"
)

type buzHashBoundaryChecker struct {
	h                     *buzhash.BuzHash
	windowSize, valueSize int
	pattern               uint32
	getBytes              getBytesFn
}

type getBytesFn func(item sequenceItem) []byte

func newBuzHashBoundaryChecker(windowSize, valueSize int, pattern uint32, getBytes getBytesFn) boundaryChecker {
	return &buzHashBoundaryChecker{buzhash.NewBuzHash(uint32(windowSize * valueSize)), windowSize, valueSize, pattern, getBytes}
}

func (b *buzHashBoundaryChecker) Write(item sequenceItem) bool {
	bytes := b.getBytes(item)
	d.Chk.Equal(b.valueSize, len(bytes))
	_, err := b.h.Write(bytes)
	d.Chk.NoError(err)
	return b.h.Sum32()&b.pattern == b.pattern
}

func (b *buzHashBoundaryChecker) WindowSize() int {
	return b.windowSize
}

File: js/src/buzhash_boundary_checker.js
// @flow

import BuzHash from './buzhash.js';
import {invariant} from './assert.js';

export type getBytesFn<T> = (item: T) => Uint8Array;

export default class BuzHashBoundaryChecker<T> {
  windowSize: number;
  valueSize: number;
  pattern: number;
  _hash: BuzHash;
  _getBytes: getBytesFn<T>;

  constructor(windowSize: number, valueSize: number, pattern: number, getBytes: getBytesFn) {
    this._hash = new BuzHash(windowSize * valueSize);
    this.windowSize = windowSize;
    this.valueSize = valueSize;
    this.pattern = pattern;
    this._getBytes = getBytes;
  }

  write(item: T): boolean {
    const bytes = this._getBytes(item);
    invariant(this.valueSize === bytes.length);
    this._hash.write(bytes);
    const sum = this._hash.sum32;
    const result = (sum & this.pattern) | 0;
    return result === this.pattern;
  }
}
